packages/event-bus-kafka/src/services/event-bus-kafka.ts

```diff
@@ -0,0 +1,121 @@
+import { Kafka, Producer, Consumer, Message } from "kafkajs"
+import { InternalModuleDeclaration, Logger } from "@medusajs/modules-sdk"
+import { EmitData } from "@medusajs/types"
+import { AbstractEventBusModuleService } from "@medusajs/utils"
+import { KafkaModuleOptions, ConsumerPayload } from "../types"
+
+type InjectedDependencies = {
+  logger: Logger
+}
+
+class KafkaEventBusService extends AbstractEventBusModuleService {
+  protected readonly logger_: Logger
+  protected readonly moduleOptions_: KafkaModuleOptions
+  // eslint-disable-next-line max-len
+  protected readonly moduleDeclaration_: InternalModuleDeclaration
+
+  protected producer: Producer
+  protected consumer: Consumer
+
+  constructor(
+    { logger }: InjectedDependencies,
+    moduleOptions: KafkaModuleOptions
+  ) {
+    // @ts-ignore
+    // eslint-disable-next-line prefer-rest-params
+    super(...arguments)
+
+    if (moduleOptions === undefined) {
+      throw new Error(`Module options are required for Kafka Relay`)
+    }
+
+    this.moduleOptions_ = moduleOptions
+    this.logger_ = logger
+
+    const kafka = new Kafka(moduleOptions.kafkaConfig)
+
+    this.producer = kafka.producer()
+    if (!moduleOptions.noConsumer) {
+      this.consumer = kafka.consumer({ groupId: "medusa-event-bus" })
+      this.initConsumer()
+    }
+  }
+
+  private getTopic(): string {
+    return this.moduleOptions_.topic ?? "medusa-event-bus"
+  }
+
+  async initConsumer() {
+    await this.consumer.connect()
+    await this.consumer.subscribe({ topic: this.getTopic() })
+    await this.consumer.run({
+      eachMessage: this.doWork,
+    })
+  }
+
+  async doWork({ message }: ConsumerPayload): Promise<void> {
+    if (!message || !message.value) {
+      return
+    }
+
+    const { eventName, data } = JSON.parse(message.value.toString())
+    const eventSubscribers = this.eventToSubscribersMap.get(eventName) || []
+    const wildcardSubscribers = this.eventToSubscribersMap.get("*") || []
+    const allSubscribers = eventSubscribers.concat(wildcardSubscribers)
+
+    this.logger_.info(
+      `Processing ${eventName} which has ${eventSubscribers.length} subscribers`
+    )
+    await Promise.all(
```

Comment:
I will have a deeper look later, but I have a question, How does the retry mechanism kick in? From what I can see if a subscriber fails, we still return a success promise.

Reply:
Same for the concurrency issues across multiple medusa instances, bullmq will claim a lock on the job when asking redis which prevent multiple instances to run the same subscriber twice across the instances. I need to see how kafka is handling this as well or do you already know?

Reply:
If you are spinning up five instances of the same consumer group, Kafka will take care of ensuring that only one of them processes the event. 

Regarding, the retry we can't really do much here - Kafka doesn't allow you to update an existing event in a Topic so we can't won't be able to keep track of which subscribers have run.